File: sdk-typescript/test/iwfcompat/state_options_override_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  StepList,
  StepMovement,
  Wait,
  goToMulti,
  gracefulComplete,
  stringCodec,
  type Context,
  type Flow,
  type Step,
  type StepDecision,
  type StepOptions,
} from "../../src/index.js";

class CompleteStep implements Step<string> {
  public readonly inputCodec = stringCodec;
  private output = "";

  public getStepType(): string {
    return "CompleteStep";
  }

  public waitFor(_context: Context, input: string): Wait {
    this.output = `${input}_state2_start`;
    throw new Error("state 2 wait failure");
  }

  public execute(context: Context, _input: string): StepDecision {
    if (!context.waitForMethodFailed()) {
      throw new Error("waitFor failure was not reported");
    }
    this.output += "_state2_decide";
    return gracefulComplete(this.output);
  }

  public getStepOptions(): StepOptions {
    return {
      waitForRetry: { maximumAttempts: 1 },
      waitForFailure: "failFlow",
    };
  }
}

class OverrideFirstStep implements Step<string> {
  public readonly inputCodec = stringCodec;
  private output = "";

  public constructor(private readonly second: CompleteStep) {}

  public getStepType(): string {
    return "OverrideFirstStep";
  }

  public waitFor(_context: Context, input: string): Wait {
    this.output = `${input}_state1_start`;
    return Wait.skipImmediately();
  }

  public execute(_context: Context, _input: string): StepDecision {
    this.output += "_state1_decide";
    return goToMulti(
      StepMovement.of(this.second, this.output, {
        waitForRetry: { maximumAttempts: 2 },
        waitForFailure: "proceed",
      }),
    );
  }
}

export class StateOptionsOverrideFlow implements Flow<string> {
  private readonly second = new CompleteStep();
  private readonly first = new OverrideFirstStep(this.second);

  public getFlowType(): string {
    return "StateOptionsOverrideFlow";
  }

  public getSteps() {
    return StepList.startStep(this.first).otherSteps(this.second);
  }
}

File: sdk-python/tests/integ/state_options_override_flow.py
# Portions of this file are derived from indeedeng/iwf-java-sdk.
# Those portions are licensed under the Apache License, Version 2.0.
# See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
#
# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications are licensed under the Super Durable Source License 1.0.
# Third-Party Materials remain under the Apache License, Version 2.0.
# See LICENSE and LEGACY_NOTICES.md.

from __future__ import annotations

from dex import (
    Context,
    Flow,
    RetryPolicy,
    Step,
    StepDecision,
    StepList,
    StepMovement,
    StepOptions,
    Wait,
    WaitForFailurePolicy,
    go_to_multi,
    graceful_complete,
)


class OverrideFirstStep(Step[str]):
    def __init__(self, second: OverrideCompleteStep) -> None:
        self.second = second
        self.output = ""

    def wait_for(self, context: Context, input: str) -> Wait:
        del context
        self.output = input + "_state1_start"
        return Wait.skip_immediately()

    def execute(self, context: Context, input: str) -> StepDecision:
        del context
        self.output += "_state1_decide"
        options = StepOptions(
            wait_for_retry=RetryPolicy(maximum_attempts=2),
            wait_for_failure=WaitForFailurePolicy.PROCEED,
        )
        return go_to_multi(StepMovement.of(self.second, self.output, options=options))


class OverrideCompleteStep(Step[str]):
    def __init__(self) -> None:
        self.output = ""

    def wait_for(self, context: Context, input: str) -> Wait:
        del context
        self.output = input + "_state2_start"
        raise RuntimeError("state 2 wait failure")

    def execute(self, context: Context, input: str) -> StepDecision:
        del input
        if not context.wait_for_method_failed():
            raise RuntimeError("wait_for failure was not reported")
        self.output += "_state2_decide"
        return graceful_complete(self.output)

    def get_step_options(self) -> StepOptions:
        return StepOptions(
            wait_for_retry=RetryPolicy(maximum_attempts=1),
            wait_for_failure=WaitForFailurePolicy.FAIL_FLOW,
        )


class StateOptionsOverrideFlow(Flow[str]):
    def __init__(self) -> None:
        self.second = OverrideCompleteStep()
        self.first = OverrideFirstStep(self.second)

    def get_steps(self) -> StepList[str]:
        return StepList.start_step(self.first).other_steps(self.second)
